### PR TITLE
fileupload network interface support

### PR DESCRIPTION
**Pull request by rajaggrawal**
Currently azure sdk does not support file upload through a specific network interface.
Adding the same as part of this PR.

Note : This PR address fix for following issue.
https://github.com/Azure/azure-iot-sdk-c/issues/1710

### CODE REVIEW

File: iothub_client/inc/internal/blob.h [hunk 1]
```diff
@@ -56,10 +56,11 @@ MU_DEFINE_ENUM_WITHOUT_INVALID(BLOB_RESULT, BLOB_RESULT_VALUES)
 * @param  httpResponse      A BUFFER_HANDLE that receives the HTTP response from the server (available only when the return value is BLOB_OK)
 * @param  certificates      A null terminated string containing CA certificates to be used
 * @param    proxyOptions    A structure that contains optional web proxy information
+* @param  curl_interface    A null terminated string containing the network interface
```
Comment (ericwol-msft) [inline iothub_client/inc/internal/blob.h:59]:
```suggestion
* @param  curl_interface    A optional null terminated string containing the network interface
```

File: iothub_client/src/iothub_client_ll_uploadtoblob.c [hunk 2]
```diff
@@ -81,6 +81,7 @@ typedef struct IOTHUB_CLIENT_LL_UPLOADTOBLOB_HANDLE_DATA_TAG
     
     char* certificates;
     HTTP_PROXY_OPTIONS http_proxy_options;
+    const char* curl_interface;
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_ll_uploadtoblob.c:84]: please move new members to end of struct

File: iothub_client/src/iothub_client_ll_uploadtoblob.c [hunk 3]
```diff
@@ -1162,6 +1171,32 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadToBlob_SetOption(IOTHUB_CLIENT_LL_UPL
             upload_data->blob_upload_timeout_secs = *(size_t*)value;
             result = IOTHUB_CLIENT_OK;
         }
+        else if (strcmp(optionName, OPTION_CURL_INTERFACE) == 0)
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_ll_uploadtoblob.c:1174]: Please submit a PR for the OPTION_CURL_INTERFACE in the Azure/azure-c-shared-utility. Then you need to update the submodule in this repo.

File: iothub_client/src/iothub_client_ll_uploadtoblob.c [hunk 4]
```diff
@@ -1162,6 +1171,32 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadToBlob_SetOption(IOTHUB_CLIENT_LL_UPL
...
+        {
+            if (value == NULL)
+            {
+                LogError("NULL is a not a valid value for curl_interface");
```
Comment (ericwol-msft) [inline iothub_client/src/iothub_client_ll_uploadtoblob.c:1178]:
```suggestion
                LogError("NULL is a not a valid value for curl interface option");
```

**Comment by jspaith** [comment]
First thanks @rajaggrawal for taking on this work.  We'd love to get this change into the SDK!

There are some changes we'll need.  My writeup below is long but given how far you've got already, I hope this is straightforward.

I've also put a few other nitpicks in the comments themselves.

**1** - c-utility change
As Eric called out, the changes to the c-utility submodule need to be brought in via a separate PR to c-utility itself.  Getting the c-utility merged up into azure-iot-c-sdk is very hard so we would do that on Microsoft side once we have th c-utility in.

I can't leave feedback directly on the c-utility changes, but in `httpapi_curl.c`  please remove `interfaceName` variable from handle and references to it.  This interfaceName is used in the SetOption call itself down to `curl_easy_setopt`, so there's no need to persist it for later.

**2** - The name `OPTION_CURL_INTERFACE` makes sense at the curl layer in HTTPAPI.  However I don't think it makes sense for clients of IoTHub device client to be calling it this.  The issue is that `OPTION_CURL_INTERFACE` makes it sound like all HTTP traffic from client would be impacted by this option.  

The way this is implemented though, only upload to blob operations actually respect this option.  So the naming of option from IoT Hub client layer should be more constrained.

I propose `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB` but am open to suggestions on the name.

What you'd need to do is to (a) add this to `iothub_client/inc/iothub_client_options.h` and (b) document in `/doc/Iothub_sdk_options.md' and of course (c) have the IoT Hub client layer .c logic use the `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB` while translating to `OPTION_CURL_INTERFACE` to HTTPAPIEx layer.

**3** - This needs unit tests around the option.  Our UT framework is home built and not the easiest to learn, but your change is pretty scoped so I can help with the places you need to add changes.
(Seriously there's potentially more lines of my writeup here than # of lines of .c files you'll need to add to the UT :)).

Also all of this is in the iothub client repo.  The c-utility doesn't really do unit tests around the platform pieces.

* Your cmake will need `-Drun_unittests=ON` set since we don't build these by default.

* For the handling of the SetOption() code, see iothub_client\tests\iothubclient_ll_u2b_ut\iothub_client_ll_u2b_ut.c
  The tests here for IoTHubClient_LL_UploadToBlob_SetOption_x509_* you should be able to copy/paste/rename option since the pattern is the same.

* For the blob side of things - step 1:
** Modify `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl`() so that it takes a "const char* networkInterface".  
   Then see where that function has the `if (proxyOptions != NULL) {...}`.  Under that block have the equivalent `if (networkInterface != NULL) {STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, proxyOptions))}`
   
   `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` is a common helper function that different test variants invoke.
   
** Once you have that plumbing, adding a `TEST_FUNCTION(Blob_UploadFromSasUri_with_network_interface)` will be a one-liner to invoke `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` and passing in a non-NULL for networkInterface.

** We also need to get some error tests where we intentionally failure inject code.  This can be difficult but your case I think it's not:).  Please just modify 
		`Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_unhappy_paths()`
			* Rename to `Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_and_network_interface_unhappy_paths()`
			* Right beneath the line to `STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, "TrustedCerts", IGNORED_PTR_ARG));`, add 
			`STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, IGNORED_PTR_ARG));`
			• That should be all, since the rest of the failure injection thing will do its magic.


**4** As I said, a long writeup but it's mostly small changes I hope that are straightforward especially as you already understand the SDK.  Don't hesitate to reach out to me if there's anything more we need to discuss.  Please tag me though since I don't always get notifications.

**Review by jspaith — CHANGES_REQUESTED** [review]
Thanks again for taking this on and we'll work together to get this in.

File: iothub_client/inc/internal/blob.h (hunk 1 above)
Comment (rajaggrawal) [reply iothub_client/inc/internal/blob.h:59]: Thanks for review @ericwol-msft , @jspaith 
I will take care of review comments.
Currently I am busy in some other priority work and  will raise a separate PR for c-utility shortly.
I need to understand UT framework to write unit Test-cases and i may need more time to write UT.

**Comment by rajaggrawal** [comment]
> First thanks @rajaggrawal for taking on this work. We'd love to get this change into the SDK!
> 
> There are some changes we'll need. My writeup below is long but given how far you've got already, I hope this is straightforward.
> 
> I've also put a few other nitpicks in the comments themselves.
> 
> **1** - c-utility change
> As Eric called out, the changes to the c-utility submodule need to be brought in via a separate PR to c-utility itself. Getting the c-utility merged up into azure-iot-c-sdk is very hard so we would do that on Microsoft side once we have th c-utility in.
> 
> I can't leave feedback directly on the c-utility changes, but in `httpapi_curl.c` please remove `interfaceName` variable from handle and references to it. This interfaceName is used in the SetOption call itself down to `curl_easy_setopt`, so there's no need to persist it for later.
> 
> **2** - The name `OPTION_CURL_INTERFACE` makes sense at the curl layer in HTTPAPI. However I don't think it makes sense for clients of IoTHub device client to be calling it this. The issue is that `OPTION_CURL_INTERFACE` makes it sound like all HTTP traffic from client would be impacted by this option.
> 
> The way this is implemented though, only upload to blob operations actually respect this option. So the naming of option from IoT Hub client layer should be more constrained.
> 
> I propose `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB` but am open to suggestions on the name.
> 
> What you'd need to do is to (a) add this to `iothub_client/inc/iothub_client_options.h` and (b) document in `/doc/Iothub_sdk_options.md' and of course (c) have the IoT Hub client layer .c logic use the `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB`while translating to`OPTION_CURL_INTERFACE` to HTTPAPIEx layer.
> 
> **3** - This needs unit tests around the option. Our UT framework is home built and not the easiest to learn, but your change is pretty scoped so I can help with the places you need to add changes.
> (Seriously there's potentially more lines of my writeup here than # of lines of .c files you'll need to add to the UT :)).
> 
> Also all of this is in the iothub client repo. The c-utility doesn't really do unit tests around the platform pieces.
> 
> * Your cmake will need `-Drun_unittests=ON` set since we don't build these by default.
> * For the handling of the SetOption() code, see iothub_client\tests\iothubclient_ll_u2b_ut\iothub_client_ll_u2b_ut.c
>   The tests here for IoTHubClient_LL_UploadToBlob_SetOption_x509_* you should be able to copy/paste/rename option since the pattern is the same.
> * For the blob side of things - step 1:
>   ** Modify `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl`() so that it takes a "const char* networkInterface".
>   Then see where that function has the `if (proxyOptions != NULL) {...}`.  Under that block have the equivalent `if (networkInterface != NULL) {STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, proxyOptions))}`
>   `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` is a common helper function that different test variants invoke.
> 
> ** Once you have that plumbing, adding a `TEST_FUNCTION(Blob_UploadFromSasUri_with_network_interface)` will be a one-liner to invoke `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` and passing in a non-NULL for networkInterface.
> 
> ** We also need to get some error tests where we intentionally failure inject code. This can be difficult but your case I think it's not:). Please just modify
> `Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_unhappy_paths()`
> * Rename to `Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_and_network_interface_unhappy_paths()`
> * Right beneath the line to `STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, "TrustedCerts", IGNORED_PTR_ARG));`, add
> `STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, IGNORED_PTR_ARG));`
> • That should be all, since the rest of the failure injection thing will do its magic.
> 
> **4** As I said, a long writeup but it's mostly small changes I hope that are straightforward especially as you already understand the SDK. Don't hesitate to reach out to me if there's anything more we need to discuss. Please tag me though since I don't always get notifications.

@jspaith  I have created a separate PR for c-utility . kindly help to merge this. 
https://github.com/Azure/azure-c-shared-utility/pull/516

**Review by jspaith — CHANGES_REQUESTED** [review]
This one is also close to being ready, with some more feedback which should be my final round unless I see other issues.  

I tried this E2E on a Linux box and confirm that "eth0" worked (or at least the options went through) but when I set a garbage interface then the call failed in curl where expected.

File: doc/Iothub_sdk_options.md [hunk 5]
```diff
@@ -140,6 +140,7 @@ The following options are supported when performing file uploads.  They are decl
 | `"x509privatekey"`           | OPTION_X509_PRIVATE_KEY         | const char*       | Sets the private key for the RSA x509 certificate
 | `"TrustedCerts"`             | OPTION_TRUSTED_CERT             | const char*       | Azure Server certificate used to validate TLS connection to IoT Hub and Azure Storage
 | `"proxy_data"`               | OPTION_HTTP_PROXY               | [HTTP_PROXY_OPTIONS*][shared-util-options-h] | Http proxy data object used for proxy connection to IoT Hub and Azure Storage
+| `"network_interface_upload_to_blob"`| OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB | const char* | Set the interface name to use as outgoing network interface for upload to blob.
```
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply doc/Iothub_sdk_options.md:143]: done.

> Please add in the comments also "NOTE: Not all HTTP clients support this option. It is currently only supported when using cURL."

done.

File: iothub_client/inc/iothub_client_options.h [hunk 6]
```diff
@@ -43,6 +43,11 @@ extern "C"
     /* DEPRECATED:: OPTION_MESSAGE_TIMEOUT is DEPRECATED! Use OPTION_SERVICE_SIDE_KEEP_ALIVE_FREQ_SECS for AMQP; MQTT has no option available. OPTION_MESSAGE_TIMEOUT legacy variable will be kept for back-compat.  */
     static STATIC_VAR_UNUSED const char* OPTION_MESSAGE_TIMEOUT = "messageTimeout";
     static STATIC_VAR_UNUSED const char* OPTION_BLOB_UPLOAD_TIMEOUT_SECS = "blob_upload_timeout_secs";
+
+    /*
+    * @brief    Set the interface name to use as outgoing network interface for upload to blob.
```
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply iothub_client/inc/iothub_client_options.h:48]: done

File: iothub_client/tests/iothubclient_ll_u2b_ut/iothub_client_ll_u2b_ut.c [hunk 7]
```diff
@@ -1316,6 +1316,27 @@ TEST_FUNCTION(IoTHubClient_LL_UploadToBlob_SetOption_x509_cert_succeeds)
     IoTHubClient_LL_UploadToBlob_Destroy(h);
 }
 
+TEST_FUNCTION(IoTHubClient_LL_UploadToBlob_SetOption_Network_Interface)
```
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply iothub_client/tests/iothubclient_ll_u2b_ut/iothub_client_ll_u2b_ut.c:1319]: done. i have fixed it now.

File: iothub_client/tests/blob_ut/blob_ut.c [hunk 8]
```diff
@@ -625,7 +625,7 @@ TEST_FUNCTION(Blob_UploadMultipleBlocksFromSasUri_when_SasUri_is_wrong_fails_2)
     ///cleanup
 }
 
-static void Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl(HTTP_PROXY_OPTIONS *proxyOptions)
+static void Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl(HTTP_PROXY_OPTIONS *proxyOptions, const char * networkInterface)
```
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply iothub_client/tests/blob_ut/blob_ut.c:628]: done. i have fixed it .

File: iothub_client/src/iothub_client_ll_uploadtoblob.c [hunk 9]
```diff
@@ -699,6 +700,10 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl(IOTHUB_CLIE
                 (void)HTTPAPIEX_SetOption(iotHubHttpApiExHandle, OPTION_CURL_VERBOSE, &curl_verbose);
             }
 
+            if (upload_data->networkInterface != NULL)
+            {
+                (void)HTTPAPIEX_SetOption(iotHubHttpApiExHandle, OPTION_CURL_INTERFACE, upload_data->networkInterface);
```
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: this changes it not needed. )HTTPAPIEX_SetOption for OPTION_CURL_INTERFACE is done by Blob_UploadMultipleBlocksFromSasUri function.

     else if ((networkInterface != NULL) && HTTPAPIEX_SetOption(httpApiExHandle, OPTION_CURL_INTERFACE, networkInterface) == HTTPAPIEX_ERROR)
                {
                    LogError("failure in setting network interface");
                    result = BLOB_ERROR;
                }

Hence removed this, change

File: iothub_client/inc/internal/blob.h (hunk 1 above)
Comment (rajaggrawal) [reply iothub_client/inc/internal/blob.h]: done

File: iothub_client/src/iothub_client_ll_uploadtoblob.c (hunk 2 above)
Comment (rajaggrawal) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: done

File: iothub_client/src/iothub_client_ll_uploadtoblob.c (hunk 3 above)
Comment (rajaggrawal) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: done.

File: iothub_client/src/iothub_client_ll_uploadtoblob.c (hunk 4 above)
Comment (rajaggrawal) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: done

**Comment by rajaggrawal** [comment]
> First thanks @rajaggrawal for taking on this work. We'd love to get this change into the SDK!
> 
> There are some changes we'll need. My writeup below is long but given how far you've got already, I hope this is straightforward.
> 
> I've also put a few other nitpicks in the comments themselves.
> 
> **1** - c-utility change
> As Eric called out, the changes to the c-utility submodule need to be brought in via a separate PR to c-utility itself. Getting the c-utility merged up into azure-iot-c-sdk is very hard so we would do that on Microsoft side once we have th c-utility in.
> 
> I can't leave feedback directly on the c-utility changes, but in `httpapi_curl.c` please remove `interfaceName` variable from handle and references to it. This interfaceName is used in the SetOption call itself down to `curl_easy_setopt`, so there's no need to persist it for later.
> 
> **2** - The name `OPTION_CURL_INTERFACE` makes sense at the curl layer in HTTPAPI. However I don't think it makes sense for clients of IoTHub device client to be calling it this. The issue is that `OPTION_CURL_INTERFACE` makes it sound like all HTTP traffic from client would be impacted by this option.
> 
> The way this is implemented though, only upload to blob operations actually respect this option. So the naming of option from IoT Hub client layer should be more constrained.
> 
> I propose `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB` but am open to suggestions on the name.
> 
> What you'd need to do is to (a) add this to `iothub_client/inc/iothub_client_options.h` and (b) document in `/doc/Iothub_sdk_options.md' and of course (c) have the IoT Hub client layer .c logic use the `OPTION_NETWORK_INTERFACE_UPLOAD_TO_BLOB`while translating to`OPTION_CURL_INTERFACE` to HTTPAPIEx layer.
> 
> **3** - This needs unit tests around the option. Our UT framework is home built and not the easiest to learn, but your change is pretty scoped so I can help with the places you need to add changes.
> (Seriously there's potentially more lines of my writeup here than # of lines of .c files you'll need to add to the UT :)).
> 
> Also all of this is in the iothub client repo. The c-utility doesn't really do unit tests around the platform pieces.
> 
> * Your cmake will need `-Drun_unittests=ON` set since we don't build these by default.
> * For the handling of the SetOption() code, see iothub_client\tests\iothubclient_ll_u2b_ut\iothub_client_ll_u2b_ut.c
>   The tests here for IoTHubClient_LL_UploadToBlob_SetOption_x509_* you should be able to copy/paste/rename option since the pattern is the same.
> * For the blob side of things - step 1:
>   ** Modify `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl`() so that it takes a "const char* networkInterface".
>   Then see where that function has the `if (proxyOptions != NULL) {...}`.  Under that block have the equivalent `if (networkInterface != NULL) {STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, proxyOptions))}`
>   `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` is a common helper function that different test variants invoke.
> 
> ** Once you have that plumbing, adding a `TEST_FUNCTION(Blob_UploadFromSasUri_with_network_interface)` will be a one-liner to invoke `Blob_UploadMultipleBlocksFromSasUri_various_sizes_happy_path_Impl` and passing in a non-NULL for networkInterface.
> 
> ** We also need to get some error tests where we intentionally failure inject code. This can be difficult but your case I think it's not:). Please just modify
> `Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_unhappy_paths()`
> * Rename to `Blob_UploadMultipleBlocksFromSasUri_64MB_with_certificate_and_network_interface_unhappy_paths()`
> * Right beneath the line to `STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, "TrustedCerts", IGNORED_PTR_ARG));`, add
> `STRICT_EXPECTED_CALL(HTTPAPIEX_SetOption(IGNORED_PTR_ARG, OPTION_CURL_INTERFACE, IGNORED_PTR_ARG));`
> • That should be all, since the rest of the failure injection thing will do its magic.
> 
> **4** As I said, a long writeup but it's mostly small changes I hope that are straightforward especially as you already understand the SDK. Don't hesitate to reach out to me if there's anything more we need to discuss. Please tag me though since I don't always get notifications.

@jspaith i have incorporated all of the comments and all the UT is passed now.  Shall we proceed to merge this PR ?

File: iothub_client/src/iothub_client_ll_uploadtoblob.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: We do actually need this call.  There are two separate HTTP client connections done by this set of functions.  The first is to Azure IoT Hub in order to get the credential information of the device and the second is a connection to Azure Storage using those credentials.  So in `Blob_UploadMultipleBlocksFromSasUri` creates its own HTTP handle around line 354 (`httpApiExHandle = HTTPAPIEX_Create(hostname);`) but this is distinct from the one created in `IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl`.  

So we'll need to set the curl network option on both handles, so the code you had originally here was correct but if that call fails then we should treat as a fatal error.

**Comment by jspaith** [comment]
@rajaggrawal  - in general changes look good except for the one around needing to set the option on both HTTP client handles, which I explained more in place.  The UT all compile now and run successfully including various Valgrind checks.

We have merged you change into c-utility.  We can't simply update the submodule from c-utility into the iot-SDK repo - yet - unfortunately.  We have other submodules we use like umqtt and they themselves use c-utility.  Our gates require all versions of c-utility to be the same Commit SHA (because having them be different could introduce horrible bugs).  (For what it's worth I don't know if we had our lives to live over again if this is how we'd make the system work, but it's what we have at moment.)

What all this means is we have some process to update our submodules the right way.  I'm somewhat optimistic that we can do this submodule work next week but I can't commit to exact dates.  There are a few other other bug fixes we must piggy back into the next submodule update which will bring your changes along.  

Sorry for delay here, but we're close.

File: iothub_client/src/iothub_client_ll_uploadtoblob.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (rajaggrawal) [reply iothub_client/src/iothub_client_ll_uploadtoblob.c]: i have added the change again and tested fileupload.

**Comment by jspaith** [comment]
@rajaggrawal - no updates at this time other than that the core of this looks good and we are now tracking the submodule update.  There should not be any additional work on your side needed for this.

**Comment by jspaith** [comment]
@rajaggrawal - sorry for delay and happy to let you know that we've update the submodules.  Could you please merge the latest master branch from C SDK into your fork and make sure the submodule fixes all look good on your side, and then tag me in the GitHub conversation when you're ready so we can take next step of getting this in.

Thanks

**Comment by rajaggrawal** [comment]
> @rajaggrawal - sorry for delay and happy to let you know that we've update the submodules. Could you please merge the latest master branch from C SDK into your fork and make sure the submodule fixes all look good on your side, and then tag me in the GitHub conversation when you're ready so we can take next step of getting this in.
> 
> Thanks

@jspaith  I have merged the latest master branch from C SDK into fork and verified. It is working fine. Let me know when the changes will be available in c-sdk ?

**Comment by jspaith** [comment]
@rajaggrawal - your code is checked in!  Once again thank you for contributing to the SDK.

**Comment by rajaggrawal** [comment]
> @rajaggrawal - your code is checked in! Once again thank you for contributing to the SDK.

@jspaith Many thanks for your time to review this PR. May i know when you planning to tag it OR  shall we take master branch as release Tag ?

**Comment by jspaith** [comment]
@rajaggrawal  - We typically only tag for our either long-term releases or intermediate ones, and not for new features or bug fixes.  So using master will be the best bet here for you all.

**Pull request merged**